### PR TITLE
Add XML documentation.

### PR DESCRIPTION
**Pull request by Choc13**

### CODE REVIEW

File: src/Symbolica.Extensions.Configuration.FSharp/Api.fs [hunk 1]
```diff
@@ -2,28 +2,62 @@ namespace Symbolica.Extensions.Configuration.FSharp
 
 [<AutoOpen>]
 module Api =
+    /// <summary>A computation expression that is used to create a binder which binds the data of config section to a user defined type.</summary>
+    /// <returns>An instance of the builder</returns>
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Api.fs:6]: You use a full stop everywhere else.

File: src/Symbolica.Extensions.Configuration.FSharp/BindResult.fs [hunk 2]
```diff
@@ -6,28 +6,61 @@ type BindResult<'a> =
...
     let apply f a : BindResult<'b> =
         match f, a with
         | Failure e1, Failure e2 -> Failure(List.append e1 e2)
         | Failure e1, Success _ -> e1 |> Failure
         | Success _, Failure e2 -> e2 |> Failure
         | Success f, Success a -> a |> f |> Success
 
+    /// <summary>Monadic bind for a <see cref="BindResult" />.</summary>
+    /// <remarks>
+    /// If the input <see cref="BindResult" /> is <c>Success</c> then the <paramref name="f" /> will be evaluated with
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/BindResult.fs:29]: You tend you use a construct like `the function <paramref name="f" />` in most places which reads slightly nicer than just `the <paramref name="f" />`.

File: src/Symbolica.Extensions.Configuration.FSharp/Binder.fs [hunk 3]
```diff
@@ -2,52 +2,150 @@ namespace Symbolica.Extensions.Configuration.FSharp
...
-
-    let run (Binder b) = b
-
-    let eval config (b: Binder<'a>) = (run b) config
-
+    /// <summary>Create a <see cref="Binder" /> from a <see cref="BindResult" />.</summary>
+    /// <remarks>
+    /// Effectively this is creating a <see cref="Binder" /> that doesn't need to read from the <see cref="IConfiguration" />.
+    /// </remarks>
+    /// <param name="bindResult">The <see cref="BindResult" /> to create this <see cref="Binder" /> from.</param>
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Binder.fs:19]: A "from which" style like you use elsewhere would probably read a bit nicer than the trailing "from".

File: src/Symbolica.Extensions.Configuration.FSharp/Binder.fs [hunk 4]
```diff
@@ -2,52 +2,150 @@ namespace Symbolica.Extensions.Configuration.FSharp
...
     let apply (f: Binder<'a -> 'b>) (a: Binder<'a>) : Binder<'b> =
         Binder(fun config -> BindResult.apply (f |> eval config) (a |> eval config))
 
+    /// <summary>Monadic bind for a <see cref="Binder" />.</summary>
+    /// <remarks>
+    /// If <paramref name="m" /> is <c>Success</c> then the <paramref name="f" /> will be evaluated with
+    /// the data contained in <paramref name="m" />; otherwise the original <c>Failure</c> will be maintained.
+    /// </remarks>
+    /// <param name="f">The function to which the value will be bound.</param>
+    /// <param name="m">The <see cref="Binder" /> that contains the input value to <paramref name="f" />.</param>
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Binder.fs:53]: Same as previously, could say `the function <paramref name="f" />`, but this one is actually less important.

File: src/Symbolica.Extensions.Configuration.FSharp/Binder.fs [hunk 5]
```diff
@@ -2,52 +2,150 @@ namespace Symbolica.Extensions.Configuration.FSharp
...
         Binder
             (fun config ->
                 m
                 |> eval config
                 |> BindResult.bind (f >> eval config))
 
+    /// <summary>Maps the inputs to the <see cref="Binder" />.</summary>
+    /// <param name="f">The function used to map the input.</param>
+    /// <param name="m">The <see cref="Binder" /> to be contramapped.</param>
+    /// <returns>A <see cref="Binder" /> whose input is mapped by <paramref name="f" />.</returns>
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Binder.fs:65]: Same here, also less important than the first one.

File: src/Symbolica.Extensions.Configuration.FSharp/Binder.fs [hunk 6]
```diff
@@ -2,52 +2,150 @@ namespace Symbolica.Extensions.Configuration.FSharp
...
     let contramap f (m: Binder<'a>) = Binder(f >> run m)
 
-    let nest (b: Binder<'a>) parentBinder =
+    /// <summary>Maps the output of the <see cref="Binder" />.</summary>
+    /// <param name="f">The function used to map the output.</param>
+    /// <param name="m">The <see cref="Binder" /> to be mapped.</param>
+    /// <returns>A <see cref="Binder" /> whose output is mapped by <paramref name="f" />.</returns>
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Binder.fs:71]: And again.

File: src/Symbolica.Extensions.Configuration.FSharp/Binder.fs [hunk 7]
```diff
@@ -59,6 +157,15 @@ module Binder =
...
                     |> Failure)
 
+    /// <summary>
+    /// Binds the optional child <see cref="IConfiguration" /> located at the <paramref name="key" /> of the
+    /// input <see cref="IConfiguration" />.
+    /// </summary>
+    /// <param name="key">The key at which to try and find a child <see cref="IConfiguration" />.</param>
+    /// <returns>
+    /// A <see cref="Binder" /> whose input is a parent <see cref="IConfiguration" /> and whose output when
+    /// evaluated is the optional chlid <see cref="IConfiguration" />.
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Binder.fs:167]: "chlid"

File: src/Symbolica.Extensions.Configuration.FSharp/Decode.fs [hunk 8]
```diff
@@ -1,25 +1,65 @@
 namespace Symbolica.Extensions.Configuration.FSharp
 
+/// <summary>
+/// A function that takes a value and returns a <see cref="Binder"/> which evaluates to <c>Success</c> if the
+/// value can be decoded otherwise <c>Failure</c>.
+/// </summary>
 type Decoder<'a> = string -> Binder<'a>
 
+/// Contains decoders for common types and combinators functions for building new decoders.
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Decode.fs:9]: Should it be "combinator functions"? You may be in French mode.

File: src/Symbolica.Extensions.Configuration.FSharp/Decode.fs [hunk 9]
```diff
@@ -1,25 +1,65 @@
...
-    let ofParser (parser: string -> bool * 'parsed) value =
-        Binder
-            (fun section ->
-                match parser value with
-                | true, x -> Success x
-                | false, _ ->
-                    Failure [ $"Could not decode '{value}' at path '{section |> path}' as type '{typeof<'parsed>}'." ])
+    /// <summary>Creates a <see cref="Decoder" /> from a System.Type.TryParse style parsing function.</summary>
+    /// <remarks>
+    /// Useful for creating a parser for a privitive type for which a <c>TryParse</c> function already exists.
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Decode.fs:23]: "privitive"

**Review by iscottb122 — CHANGES_REQUESTED** [review]
Required spelling and optional grammar.

File: src/Symbolica.Extensions.Configuration.FSharp/BindResult.fs (hunk 2 above)
Comment (Choc13) [reply src/Symbolica.Extensions.Configuration.FSharp/BindResult.fs:29]: Yeah this seems like a typo here. Good spot.

File: src/Symbolica.Extensions.Configuration.FSharp/Decode.fs (hunk 8 above)
Comment (Choc13) [reply src/Symbolica.Extensions.Configuration.FSharp/Decode.fs:9]: It should. Nice spot.

File: src/Symbolica.Extensions.Configuration.FSharp/BindResult.fs [hunk 10]
```diff
@@ -6,28 +6,61 @@ type BindResult<'a> =
...
+    /// in the <c>Failure</c> case.
+    /// </summary>
     let defaultWith f =
         function
         | Success x -> x
         | Failure e -> e |> f
 
+    /// <summary>Maps the output of the <see cref="BindResult" />.</summary>
+    /// <param name="f">The function used to map the output.</param>
+    /// <returns>A <see cref="BindResult" /> whose output is mapped by <paramref name="f" />.</returns>
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/BindResult.fs:50]: This one could possibly also read better as `the function <paramref name="f" />`.

File: src/Symbolica.Extensions.Configuration.FSharp/Binder.fs [hunk 11]
```diff
@@ -2,52 +2,150 @@ namespace Symbolica.Extensions.Configuration.FSharp
...
+    /// </remarks>
+    /// <param name="f">The function to which the value will be applied.</param>
+    /// <param name="a">The value to be applied to the function.</param>
+    /// <returns>A <see cref="Binder" /> containing the result of applying the value to the function.</returns>
     let apply (f: Binder<'a -> 'b>) (a: Binder<'a>) : Binder<'b> =
         Binder(fun config -> BindResult.apply (f |> eval config) (a |> eval config))
 
+    /// <summary>Monadic bind for a <see cref="Binder" />.</summary>
+    /// <remarks>
+    /// If <paramref name="m" /> is <c>Success</c> then the <paramref name="f" /> will be evaluated with
```
Comment (iscottb122) [inline src/Symbolica.Extensions.Configuration.FSharp/Binder.fs:49]: And this one definitely would.

**Review by iscottb122 — CHANGES_REQUESTED** [review]
Sorry, missed a couple first time round.

**Review by iscottb122 — APPROVED** [review]

**Pull request merged**